### ISSUE TITLE
Total charge of organization is bigger than the sum of charges for each space

### ISSUE THREAD

**Issue by georgethebeatle** [cloudfoundry-incubator/cf-abacus#63]
I have been playing with the `cf-bridge` and the report that it generates off of a local Cloud Foundry instance and stumbled upon some strange behaviour. 

I have one organization containing two spaces each of which has one application running. One app has one instance and the other has two. The bridge is reporting the `linux-container` resource that comes with Abacus. What I notices is that in the final report the sum of the charges for each space is less than the total charge for the entire organization.

As the `linux-container` metric is somewhat complex I decided to simplify the formulae so that it is easier to track what is going on. Here is what I ended up with:

```javascript
    meter: (function (m) {
      return {
        consuming: m.running_instances,
        since: m.start
      };
    }).toString(),

    accumulate: (function (a, qty) {
      return qty;
    }).toString(),

    aggregate: (function (a, prev, curr) {
      return a ? {
        consumed: a.consumed + curr.consuming,
        consuming: curr.consuming,
        since: curr.since
      } : {
        consumed: 0,
        consuming: curr.consuming,
        since: curr.since
      };
    }).toString(),

    rate: (function (price, qty) {
      return {
        burned: qty.consumed * 1,
        burning: qty.consuming * 1,
        since: qty.since
      };
    }).toString(),

    summarize: (function (t, qty) {
      return qty.consumed;
    }).toString(),

    charge: (function (t, cost) {
      return cost.burned;
    }).toString()
```

Basically adding up the total number of instances in `consumed`, keeping track of current number of instances in `consuming` and then charging only for `consumed` instances.

Here is the report I got:

```json
{
  "organization_id": "91a2e8bc-1d6d-4e99-84bf-0b78b664db75",
  "start": 1443744000000,
  "end": 1443830399999,
  "resources": [
    {
      "resource_id": "linux-container",
      "plans": [
        {
          "plan_id": "basic",
          "aggregated_usage": [
            {
              "metric": "memory",
              "quantity": {
                "consumed": 1,
                "consuming": 1,
                "since": 1443784064000
              },
              "cost": {
                "burned": 1,
                "burning": 1,
                "since": 1443784064000
              },
              "summary": 1,
              "charge": 1
            }
          ],
          "charge": 1
        }
      ],
      "aggregated_usage": [
        {
          "metric": "memory",
          "quantity": {
            "consumed": 1,
            "consuming": 1,
            "since": 1443784064000
          },
          "summary": 1,
          "charge": 1
        }
      ],
      "charge": 1
    }
  ],
  "spaces": [
    {
      "space_id": "cf639c05-b275-48ae-8b1f-b56ff84d53c6",
      "resources": [
        {
          "resource_id": "linux-container",
          "plans": [
            {
              "plan_id": "basic",
              "aggregated_usage": [
                {
                  "metric": "memory",
                  "quantity": {
                    "consumed": 0,
                    "consuming": 2,
                    "since": 1443784018000
                  },
                  "cost": {
                    "burned": 0,
                    "burning": 2,
                    "since": 1443784018000
                  },
                  "summary": 0,
                  "charge": 0
                }
              ],
              "charge": 0
            }
          ],
          "aggregated_usage": [
            {
              "metric": "memory",
              "quantity": {
                "consumed": 0,
                "consuming": 2,
                "since": 1443784018000
              },
              "summary": 0,
              "charge": 0
            }
          ],
          "charge": 0
        }
      ],
      "consumers": [
        {
          "consumer_id": "ALL",
          "resources": [
            {
              "resource_id": "linux-container",
              "plans": [
                {
                  "plan_id": "basic",
                  "aggregated_usage": [
                    {
                      "metric": "memory",
                      "quantity": {
                        "consumed": 0,
                        "consuming": 2,
                        "since": 1443784018000
                      },
                      "cost": {
                        "burned": 0,
                        "burning": 2,
                        "since": 1443784018000
                      },
                      "summary": 0,
                      "charge": 0
                    }
                  ],
                  "charge": 0
                }
              ],
              "aggregated_usage": [
                {
                  "metric": "memory",
                  "quantity": {
                    "consumed": 0,
                    "consuming": 2,
                    "since": 1443784018000
                  },
                  "summary": 0,
                  "charge": 0
                }
              ],
              "charge": 0
            }
          ],
          "charge": 0
        }
      ],
      "charge": 0
    },
    {
      "space_id": "20cf3eb5-6750-4929-83d2-b1c6d80fd35a",
      "resources": [
        {
          "resource_id": "linux-container",
          "plans": [
            {
              "plan_id": "basic",
              "aggregated_usage": [
                {
                  "metric": "memory",
                  "quantity": {
                    "consumed": 0,
                    "consuming": 1,
                    "since": 1443784064000
                  },
                  "cost": {
                    "burned": 0,
                    "burning": 1,
                    "since": 1443784064000
                  },
                  "summary": 0,
                  "charge": 0
                }
              ],
              "charge": 0
            }
          ],
          "aggregated_usage": [
            {
              "metric": "memory",
              "quantity": {
                "consumed": 0,
                "consuming": 1,
                "since": 1443784064000
              },
              "summary": 0,
              "charge": 0
            }
          ],
          "charge": 0
        }
      ],
      "consumers": [
        {
          "consumer_id": "ALL",
          "resources": [
            {
              "resource_id": "linux-container",
              "plans": [
                {
                  "plan_id": "basic",
                  "aggregated_usage": [
                    {
                      "metric": "memory",
                      "quantity": {
                        "consumed": 0,
                        "consuming": 1,
                        "since": 1443784064000
                      },
                      "cost": {
                        "burned": 0,
                        "burning": 1,
                        "since": 1443784064000
                      },
                      "summary": 0,
                      "charge": 0
                    }
                  ],
                  "charge": 0
                }
              ],
              "aggregated_usage": [
                {
                  "metric": "memory",
                  "quantity": {
                    "consumed": 0,
                    "consuming": 1,
                    "since": 1443784064000
                  },
                  "summary": 0,
                  "charge": 0
                }
              ],
              "charge": 0
            }
          ],
          "charge": 0
        }
      ],
      "charge": 0
    }
  ],
  "id": "k/91a2e8bc-1d6d-4e99-84bf-0b78b664db75/t/0001443744000000",
  "charge": 1
}
```

In the per-space aggregation I am getting the expected `{consumed: 0, consuming: x}` where x is the number of instances for the app, but in the per-resource aggregation I get `{consumed: 1, consuming: 1}` and I do not understand why.

Apart from that it seems that the total charge = total charge for resources + total charge for spaces which looks like a bug. After all these resources are used by the same spaces so we have double counting. Or maybe I don't get something. Thoughts?

**Comment by jsdelfino** [comment]
IIUC the expected result should be total charge for the org == total of the charges for the spaces == total charge for the resources. You are correct that the charges for the spaces should not added to the charges for the resources, so there must be a bug in the aggregation or rating code.

**Comment by jsdelfino** [comment]
Ben, now that the time window support is in place can you take a look at this? Thanks!

**Comment by BetaFood** [comment]
The charges don't add up due to the functions defined in the resource configuration. There was an issue in the sample linux-container resource config as well in the charge and summary functions. That has been addressed by 4c2ac1ec3c95f2b8ad16fbabd4068c93727d1b92. These two functions were trying to divide by hour in milliseconds, but they ended up only dividing the real-time calculation portion and did not touch the consumed/burned values that were from the document leading to an addition on two different timescales.

In terms of the functions that have been provided in this issue, there's a couple of things I think I can point out.
Simplifying the aggregation function to not use the time can cause the consumed value to not be properly represented by the time of the usage submissions. This can affect the charge & summarize functions if they include calculations based upon the time of report.
Charge & summarize do not take into account the time different between the the last submitted usage and the time of the report. Since that's the case, any immediate charges in the space won't be calculated.

In aggregation for this particular case, the consumed value will always be 0 on the first accumulated usage. In the spaces, I'm assuming that there is only one accumulated usage per space. When it gets to the organization level, that is no longer the case and consumed will be potentially calculated beyond 0(in this case, it's 1). Since rate (and subsequently charge) is based off of just the consumed value, the spaces that aggregated to 0 consume will have 0 charge while the organization is returning 1.

Changing the aggregate/charge/time functions to be wary of the time should make it so the quantity/costs/charge/summary match up on the space and organization levels.

Does that help in any way?

**Comment by jsdelfino** [comment]
Tested again on v0.0.2-rc.1 and the charges look good to me now. Closing, please reopen if needed.
